R/tax_unique.R

```diff
@@ -0,0 +1,355 @@
+  if (!is.null(binomial)) {
+    occdf$binomial <- gsub("_", " ", occdf$binomial)
+  }
+
+  if ((!is.null(class) && any(grepl("[[:punct:]]", occdf$class))) ||
+      (!is.null(order) && any(grepl("[[:punct:]]", occdf$order))) ||
+      (!is.null(family) && any(grepl("[[:punct:]]", occdf$family))) ||
+      (!is.null(genus) && any(grepl("[[:punct:]]", occdf$genus))) ||
+      (!is.null(species) && any(grepl("[[:punct:]]", occdf$species))) ||
+      (!is.null(binomial) && any(grepl("[^[:alnum:][:space:]]",
+                                       occdf$binomial))) ||
+      (!is.null(names) && any(grepl("[[:punct:]]", occdf$names)))) {
+    stop("Name columns should not contain punctuation")
+  }
+
+  if ((resolution == "species") && (is.null(binomial)) && (is.null(species))
+      && (is.null(names))) {
+    stop("Species names must be supplied by specifying 'binomial', 'genus' and
+    'species', or 'genus' and 'names' columns to estimate diversity at species
+    level")
+  }
+
+  if ((resolution == "genus") && (is.null(binomial)) && (is.null(genus))
+      && (is.null(names))) {
+    stop("Genus names must be supplied by specifying 'binomial' or 'genus'
+    columns to estimate diversity at genus level")
+  }
+
+  if ((resolution != "species") && (resolution != "genus")) {
+    stop("Resolution must be 'species' or 'genus'")
+  }
+
+#Run function
+  genus_species <- NULL
+  occurrences <- subset(occdf, select = c(binomial, species, genus, family,
+                                          order, class, names))
+
+  if (!is.null(names)) {
+    #If name is not a binomial, replace with NA
+    occurrences$names[grep(" ", occurrences$names, invert = TRUE)] <- NA
+
+    #Rename column
+    colnames(occurrences)[colnames(occurrences) == "names"] <-
+      "genus_species"
+  }
+
+  if (!is.null(binomial)) {
+    #Rename column
+    colnames(occurrences)[colnames(occurrences) == "binomial"] <-
+      "genus_species"
+
+    #If no 'genus' column is given, make one
+    if (is.null(genus)) {
+      occurrences$genus <- sub(" .*", "", occurrences$genus_species)
+    }
+  }
+
+  if (!is.null(species)) {
+    #Create genus_species column
+    occurrences$genus_species <- paste(occurrences$genus, occurrences$species)
+    occurrences <- subset(occurrences, select = -c(species))
+
+    #If one or both of the values was NA, set to NA
+    occurrences$genus_species[grep("NA", occurrences$genus_species)] <- NA
+  }
+
+  #Remove absolute repeats
+  occurrences <- unique(occurrences)
+
+  #Create an empty dataset to collect unique occurrences in
+  to_retain <- data.frame()
+
+  if (resolution == "species") {
+    #Retain occurrences identified to species level and remove from dataframe
+    to_retain <- rbind(to_retain, occurrences[!is.na(occurrences$genus_species),
+                                              ])
+    occurrences <- occurrences[is.na(occurrences$genus_species), ]
+
+    #Retain occurrences identified to genus level and not already in dataset
```

Comment:
Yes, this could definitely happen, especially for occdfs including taxa from lots of different clades. But as you say, the scale of the problem is likely to be quite small (proportional to the volume of data). I think this would actually be quite fiddly to implement, so I would personally prefer to leave this for now and flag it as something to fix in future. There is a warning about homonyms in the documentation, so we are being honest about it. What do you think?

Reply:
Happy with that, maybe log an issue to serve as a reminder for us!